Shaded Jackson

We are implementing a neural network for one of our products using the deep learning libraries  for java. One of the dependencies for deeplearning4j-nn.jar is shaded Jackson by nd4j. This is blowing up in the OWASP security scan of the libraries. Therefore, we cannot use this dependency. Is there an alternate option around the shaded jar?

Below are  all the jars we are using for this project. This is the smallest list of dependencies we could find to get the model running.  I would like to know if we can reduce this further down and/or reduce the size of some of the big ones.? Please guide me.


#### Version Information
deeplearning4j-nn-1.0.0-beta5.jar;
guava-1.0.0-beta5.jar;
jackson-1.0.0-beta5.jar;
javacpp-1.5.1-1.jar;
mkl-dnn-0.20-1.5.1-linux-x86_64.jar;
mkl-dnn-0.20-1.5.1-windows-x86_64.jar;
nd4j-api-1.0.0-beta5.jar;
nd4j-buffer-1.0.0-beta5.jar;
nd4j-common-1.0.0-beta5.jar;
nd4j-context-1.0.0-beta5.jar;
nd4j-native-1.0.0-beta5.jar;
nd4j-native-1.0.0-beta5-linux-x86_64.jar;
nd4j-native-1.0.0-beta5-windows-x86_64.jar;
nd4j-native-api-1.0.0-beta5.jar;
nd4j-native-platform-1.0.0-beta5.jar;
openblas-0.3.6-1.5.1.jar;
openblas-0.3.6-1.5.1-linux-x86_64.jar;
openblas-0.3.6-1.5.1-windows-x86_64.jar;

I get an error without it

Exception in thread "pool-2-thread-1" java.lang.ExceptionInInitializerError
Caused by: java.lang.UnsatisfiedLinkError: no jnind4jcpu in
java.library.path
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
at java.lang.Runtime.loadLibrary0(Runtime.java:870)
at java.lang.System.loadLibrary(System.java:1122)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1487)
at org.bytedeco.javacpp.Loader.load(Loader.java:1190)
at org.bytedeco.javacpp.Loader.load(Loader.java:1040)
at org.nd4j.nativeblas.Nd4jCpu.<clinit>(Nd4jCpu.java:10)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:264)
at org.nd4j.nativeblas.NativeOpsHolder.<init>(NativeOpsHolder.java:84)
... 16 more
Caused by: java.lang.UnsatisfiedLinkError:
C:\Users\svarghese\.javacpp\cache\nd4j-native-windows-x86_64.jar\org\nd4j\nativeblas\windows-x86_64\jnind4jcpu.dll:
Can't find dependent libraries
at java.lang.ClassLoader$NativeLibrary.load(Native Method)
at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
at java.lang.Runtime.load0(Runtime.java:809)
at java.lang.System.load(System.java:1086)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1461)
... 22 more

On Tue, Nov 26, 2019 at 8:18 AM raver119 <notifications@github.com> wrote:

> mkl-dnn shouldn't be needed in beta5 and subsequent releases, since it's
> compiled into libnd4j staticallym and comes in
> nd4j-native-windows-x86_64.jar. OpenBLAS or MKL is required though.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/deeplearning4j/issues/8437?email_source=notifications&email_token=ANOWJNZWBPQFHWJRWWDLZ2DQVUO3FA5CNFSM4JQSXLJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFF622A#issuecomment-558624104>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANOWJN64B44QZWIOX26X7WLQVUO3FANCNFSM4JQSXLJA>
> .
>

--- a/nd4j/nd4j-shade/jackson/pom.xml
+++ b/nd4j/nd4j-shade/jackson/pom.xml
@@ -240,6 +240,17 @@
                         </relocation>
                     </relocations>
 
+                    <filters>
+                        <filter>
+                            <artifact>*:*</artifact>
+                            <excludes>
+                                <exclude>META-INF/services/javax.xml.stream.XMLOutputFactory</exclude>
+                                <exclude>META-INF/services/javax.xml.stream.XMLInputFactory</exclude>
+                                <exclude>META-INF/services/javax.xml.stream.XMLEventFactory</exclude>
+                            </excludes>
+                        </filter>
+                    </filters>
+
                 </configuration>
             </plugin>
             <plugin>

